Add LooseVersion class with risky comparison, deprecate parse_loose_version

LeonGr opened the pull request:
Relates to #9489 

Replace the type-unsafe `certbot.util.parse_loose_version` with class `certbot.util.LooseVersion`, which has type-safe comparison.
`parse_loose_version` would throw a `TypeError` when two version would have two differently typed version components in the same position (e.g. `2.0` and `2a`). The `LooseVersion` class has type-safe comparison method with an internal 'incomparable' state. This will cause inequality comparisons to always return `False`. For example both `LooseVersion('2.0') > LooseVersion('2.a')` and `LooseVersion('2.0') < LooseVersion('2.a')` return `False`.

I'm open for feedback on this approach, as it still has the potential for misuse, although it won't crash.

Edit: discussed [here on Mattermost](https://opensource.eff.org/eff-open-source/pl/khc1wx9dm3rgmj5h3ok3nkb38e).

## Pull Request Checklist

- [x] The Certbot team has recently expressed interest in reviewing a PR for this. If not, this PR may be closed due our limited resources and need to prioritize how we spend them.
- [x] If the change being made is to a [distributed component](https://certbot.eff.org/docs/contributing.html#code-components-and-layout), edit the `master` section of `certbot/CHANGELOG.md` to include a description of the change being made.
- [x] Add or update any documentation as needed to support the changes in this PR.
- [x] Include your name in `AUTHORS.md` if you like.

LeonGr commented on certbot/certbot/util.py:
```diff
@@ -48,6 +49,154 @@ class CSR(NamedTuple):
+                             other.loose_version)
+                return self.Comparison.INCOMPARIBLE
+
+            if isinstance(version_component_other, str) and \
+               isinstance(version_component_self, str):
+                if version_component_self > version_component_other:
+                    return self.Comparison.GREATER
+
+                if version_component_self < version_component_other:
+                    return self.Comparison.LESS
+
+            elif isinstance(version_component_other, int) and \
+                 isinstance(version_component_self, int):
+                if version_component_self > version_component_other:
+                    return self.Comparison.GREATER
+
+                if version_component_self < version_component_other:
+                    return self.Comparison.LESS
```
I'm not very familiar with mypy, is there a better way to do this?

bmw commented:
Thanks for the PR!

I'm currently working through a bit of a backlog of work to do and review, but I plan to get to this myself soon unless someone else on the Certbot team happens to have the time and interest before I do.

LeonGr commented:
Thanks for the review. Your proposed approach sounds reasonable. Even though it isn't crash proof, the cause would be many times more obvious than the `TypeError: '<' not supported between instances of 'str' and 'int'` exception.
I'm not the biggest fan of returning a negative, zero, or positive values, as it is not immediately obvious what these indicate. However, it makes sense to follow the convention.
I will try to get to work on it soon.

LeonGr commented:
I implemented your feedback. I also added some helper functions such as `self.try_risky_less_equal(other)`, because it was quite clunky to have to write `self.try_risky_comparison(other) in (0, -1)` for a simple `self <= other` comparison. What do you think?

bmw commented on certbot/certbot/util.py:
```diff
@@ -48,6 +50,200 @@ class CSR(NamedTuple):
+    def __init__(self, version_string: str) -> None:
+        self.version_components = self.__parse_loose_version(version_string)
+
+    def __parse_loose_version(self, version_string: str) -> List[Union[int, str]]:
```
nit: We generally just prefix things with a single underscore in this codebase.

bmw commented on certbot/certbot/util.py:
```diff
@@ -48,6 +50,200 @@ class CSR(NamedTuple):
+                pass
+        return components
+
+    class Comparison(Enum):
```
I think the rest of this class can probably be simplified. If the type annotation on the comparison function is also `LooseVersion`, we arguably don't have to check for any other type. After doing that, I think the rest of the comparison function could be as simple as:
```
return (self.version_components > other.version_components) - (self.version_components < other.version_components)
```
See https://docs.python.org/3.0/whatsnew/3.0.html#ordering-comparisons which should probably be referenced in a code comment since that's pretty magical.

Maybe we'd want to catch and reraise the `TypeError` with a different error message, but that otherwise seems like it should be sufficient to me. What do you think?

LeonGr replied:
This would simplify the implementation considerably, but this would give a slightly different result than the current implementation. "2.0" and "2.0.0" would currently be equal, but the simplified version "2.0.0" > "2.0". If this doesn't matter I think we should go with the simplified version.

LeonGr replied:
Sorry for the late response, I have been quite busy lately. 
Your proposed approach passes all my tests :+1: 
I'll implement it with a `ValueError` reraise.

bmw commented on certbot/certbot/util.py:
```diff
@@ -48,6 +50,200 @@ class CSR(NamedTuple):
+                             "due to comparison of version components with different types."
+                             .format(self.version_components, other.version_components))
+
+    # Provide helper methods for comparison without having to use the returned integer value.
```
I don't think we should provide these. While it's not common in modern Python, a function like this is fairly common in other languages. See https://en.wikipedia.org/wiki/Three-way_comparison. You can check for what you want by using the `<=`, `<`, `>`, and `>=` operators with the return value of the function and `0`. See https://github.com/python/cpython/blob/579e985e75fd682e314c7b0cce6b1751abe8a7d0/Lib/functools.py#L212-L221.

bmw commented on certbot/certbot/util.py:
```diff
@@ -48,6 +50,200 @@ class CSR(NamedTuple):
+        """
+        return self.try_risky_comparison(other) in (0, 1)
+
+    # Prevent the use of regular comparison operators
```
I also don't think we should provide these. With them, `mypy` doesn't complain about trying to compare two `LooseVersion` objects and without them it does.

bmw commented on certbot-apache/certbot_apache/_internal/configurator.py:
```diff
@@ -162,10 +162,10 @@ def pick_apache_config(self, warn_on_no_mod_ssl: bool = True) -> str:
         openssl_version = self.openssl_version(warn_on_no_mod_ssl)
         if self.version < (2, 4, 11) or not openssl_version or \
-            util.parse_loose_version(openssl_version) < min_openssl_version:
+            util.LooseVersion(openssl_version).try_risky_less(min_openssl_version):
```
I think that with the approach taken in this PR each use of the class should probably handle the case where the objects are incomparable unless we're either certain that won't happen or are OK with a `TypeError` being raised out of the function using `LooseVersion`.

If you want to try and make judgement calls about what we should do in each case, feel free, but otherwise maybe don't switch (all) callers over to the new comparison function and leave `parse_loose_version` undeprecated for now. Switching over the rest and deprecating the function could be saved for future work.

bmw commented on certbot/certbot/util.py:
```diff
@@ -48,6 +49,82 @@ class CSR(NamedTuple):
+
+        self.version_components = components
+
+    def try_risky_comparison(self, other: Any) -> int:
```
I think this change helps by giving mypy errors (in addition to the runtime errors) if this function is called with the wrong type.
```suggestion
    def try_risky_comparison(self, other: 'LooseVersion') -> int:
```

bmw reviewed (changes requested):
Sorry for the late response myself! I was on vacation and then busy catching up with things.

Other than my small inline comments, this LGTM!

As a reminder for myself or whoever merges this, I think we should keep https://github.com/certbot/certbot/issues/9489 open and/or open a new issue to track switching everything to the new `LooseVersion` class and deprecating the old function.

LeonGr commented on letstest/scripts/test_openssl_version.py:
```diff
@@ -13,8 +13,8 @@ def main(openssl_version, apache_version):
         contents = f.read()
-    if util.parse_loose_version(apache_version.strip()) < util.parse_loose_version('2.4.11') or \
-        util.parse_loose_version(openssl_version.strip()) < util.parse_loose_version('1.0.2l'):
+    if util.LooseVersion(apache_version.strip()).test_risky_less(util.LooseVersion('2.4.11')) or \
+        util.LooseVersion(openssl_version.strip()).test_risky_less(util.LooseVersion('1.0.2l')):
```
Whoops, I thought I reverted all of them

bmw reviewed (approved):
LGTM! Thanks for picking this up and brainstorming approaches with me.
